src/GovUk.Education.ExploreEducationStatistics.Data.Services/ViewModels/LocationViewModelBuilder.cs

```diff
@@ -0,0 +1,123 @@
+        Dictionary<GeographicLevel, Dictionary<string, GeoJson>>? geoJson = null)
+    {
+        var locationAttributes = locations.GetLocationAttributesHierarchical(hierarchies);
+        return locationAttributes.ToDictionary(
+            pair => pair.Key,
+            pair =>
+            {
+                var geoJsonByCode = geoJson?.GetValueOrDefault(pair.Key);
+                return BuildLocationAttributeViewModels(pair.Value, geoJsonByCode);
+            });
+    }
+
+    private static List<LocationAttributeViewModel> BuildLocationAttributeViewModels(
+        IEnumerable<LocationAttributeNode> locationAttributes,
+        IReadOnlyDictionary<string, GeoJson>? geoJsonByCode)
+    {
+        return DeduplicateLocationViewModels(
+            locationAttributes
+                .OrderBy(OrderLocationAttributes)
+                .Select(locationAttribute => BuildLocationAttributeViewModel(locationAttribute, geoJsonByCode))
+                .ToList()
+        );
+    }
+
+    private static LocationAttributeViewModel BuildLocationAttributeViewModel(
+        LocationAttributeNode locationAttributeNode,
+        IReadOnlyDictionary<string, GeoJson>? geoJsonByCode)
+    {
+        var locationAttribute = locationAttributeNode.Attribute;
+        return locationAttributeNode.IsLeaf
+            ? BuildLocationAttributeLeafViewModel(locationAttributeNode, geoJsonByCode)
+            : new LocationAttributeViewModel
+            {
+                Label = locationAttribute.Name ?? string.Empty,
+                Level = locationAttribute.GeographicLevel,
+                Value = locationAttribute.GetCodeOrFallback(),
+                Options = BuildLocationAttributeViewModels(locationAttributeNode.Children, geoJsonByCode)
+            };
+    }
+
+    private static LocationAttributeViewModel BuildLocationAttributeLeafViewModel(
+        LocationAttributeNode locationAttributeNode,
+        IReadOnlyDictionary<string, GeoJson>? geoJsonByCode)
+    {
+        var locationAttribute = locationAttributeNode.Attribute;
+        var code = locationAttribute.GetCodeOrFallback();
+
+        var geoJson = code.IsNullOrEmpty()
+            ? null
+            : geoJsonByCode?.GetValueOrDefault(code)?.Deserialized;
+
+        return new LocationAttributeViewModel
+        {
+            Id = locationAttributeNode.LocationId,
+            GeoJson = geoJson,
+            Label = locationAttribute.Name ?? string.Empty,
+            Value = code
+        };
+    }
+
+    private static List<LocationAttributeViewModel> DeduplicateLocationViewModels(
+        List<LocationAttributeViewModel> locations)
+    {
+        // Find duplicates by label which also have the same value.
+        // They are unique by some other attribute of location and appending the value of this attribute won't deduplicate them.
+        // TODO EES-2954 enhance the deduplication to append other attributes of the location
+        var notDistinctByLabelAndValue = locations
```

Comment:
So at the moment after this work goes in, we'll still have the duplicate Locations in the front end that are appearing because of some other hidden attribute, but we'll not see the empty braces where empty value was being appended anymore?  And 2954 will address the hidden attribute surfacing?

Reply:
Yep, that's the plan if we decide to do it. There's some considerations added to the comments of EES-2954 and a plan for implementing it.

With this refactoring if should be easier to do now as we have the 'full' locations closer to where we do the deduplication.

We could also move the point where the deduplication happens to the area where the narrowing from Location to LocationAttribute occurs when building the location hierarchy rather than it operating on view models as it does at present.